src/rust/plugin-work-queue/src/server.rs

```diff
@@ -24,111 +31,266 @@ use rust_proto::plugin_work_queue::{
+};
 use tonic::{
     transport::Server,
     Request,
     Response,
     Status,
 };
 
+use crate::{
+    psql_queue::{
+        PsqlQueue,
+        PsqlQueueError,
+    },
+    PluginWorkQueueServiceConfig,
+};
+
 #[derive(Debug, thiserror::Error)]
-pub enum PluginWorkQueueError {}
+pub enum PluginWorkQueueError {
+    #[error("PsqlQueueError")]
+    PsqlQueueError(#[from] PsqlQueueError),
+    #[error("From<PluginWorkQueueDeserializationError>")]
+    DeserializationError(#[from] PluginWorkQueueDeserializationError),
+}
 
-pub struct PluginWorkQueue {}
+impl From<PluginWorkQueueError> for Status {
+    fn from(err: PluginWorkQueueError) -> Self {
+        match err {
+            PluginWorkQueueError::PsqlQueueError(_) => Status::internal("Sql Error"),
+            PluginWorkQueueError::DeserializationError(_) => {
+                Status::invalid_argument("Invalid argument")
+            }
+        }
+    }
+}
+
+pub struct PluginWorkQueue {
+    pub(crate) queue: PsqlQueue,
+}
+
+impl From<PsqlQueue> for PluginWorkQueue {
+    fn from(queue: PsqlQueue) -> Self {
+        Self { queue }
+    }
+}
+
+impl From<Pool<Postgres>> for PluginWorkQueue {
+    fn from(pool: Pool<Postgres>) -> Self {
+        Self {
+            queue: PsqlQueue { pool },
+        }
+    }
+}
 
 impl PluginWorkQueue {
     #[allow(dead_code)]
     async fn put_execute_generator(
         &self,
-        _request: PutExecuteGeneratorRequest,
+        request: PutExecuteGeneratorRequest,
     ) -> Result<PutExecuteGeneratorResponse, PluginWorkQueueError> {
-        todo!()
+        let tenant_id = request.execution_job.tenant_id;
+        let plugin_id = request.execution_job.plugin_id;
+        let data = request.execution_job.data;
+
+        self.queue
+            .put_generator_message(plugin_id, data, tenant_id)
+            .await?;
+
+        Ok(PutExecuteGeneratorResponse {})
     }
 
     #[allow(dead_code)]
     async fn put_execute_analyzer(
         &self,
-        _request: PutExecuteAnalyzerRequest,
+        request: PutExecuteAnalyzerRequest,
     ) -> Result<PutExecuteAnalyzerResponse, PluginWorkQueueError> {
-        todo!()
+        let tenant_id = request.execution_job.tenant_id;
+        let plugin_id = request.execution_job.plugin_id;
+        let data = request.execution_job.data;
+
+        self.queue
+            .put_analyzer_message(plugin_id, data, tenant_id)
+            .await?;
+
+        Ok(PutExecuteAnalyzerResponse {})
     }
 
     #[allow(dead_code)]
     async fn get_execute_generator(
         &self,
         _request: GetExecuteGeneratorRequest,
     ) -> Result<GetExecuteGeneratorResponse, PluginWorkQueueError> {
-        todo!()
+        let message = self.queue.get_analyzer_message().await?;
+        let message = match message {
+            Some(message) => message,
+            None => {
+                return Ok(GetExecuteGeneratorResponse {
+                    execution_job: None,
+                    request_id: 0,
```

Comment:
Is this not a NoAvailableJobs scenario? 

Even if it's not: it's not immediately obvious to me _what_ scenario this is.

Reply:
This is a NoAvailableJobs scenario. it's not really exposed on the rust end other than as an Optional. I could bring it over to the rust side.